internal/pkg/agent/application/coordinator/coordinator.go

```diff
@@ -194,7 +197,7 @@ type Coordinator struct {
 	// public accessors like State(), SetLogLevel(), etc.
 	state             State
 	stateBroadcaster  *broadcaster.Broadcaster[State]
-	stateNeedsRefresh bool
+	stateNeedsRefresh atomic.Bool
```

Comment:
This change was needed to prevent a data race like so:

```
[2023-10-13T13:08:03.453Z] === FAIL: internal/pkg/agent/application/actions/handlers TestUpgradeHandler (2.21s)
[2023-10-13T13:08:03.453Z] ==================
[2023-10-13T13:08:03.453Z] WARNING: DATA RACE
[2023-10-13T13:08:03.453Z] Write at 0x00c000952cb8 by goroutine 115:
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).setOverrideState()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator_state.go:118 +0x704
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).runLoopIteration()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator.go:892 +0x6c0
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).runner()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator.go:854 +0x49c
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).Run()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator.go:613 +0x330
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers.TestUpgradeHandler.func1()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers/handler_action_upgrade_test.go:75 +0x4c
[2023-10-13T13:08:03.453Z] 
[2023-10-13T13:08:03.453Z] Previous write at 0x00c000952cb8 by goroutine 117:
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).setUpgradeDetails()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator_state.go:246 +0x78
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).setUpgradeDetails-fm()
[2023-10-13T13:08:03.453Z]       <autogenerated>:1 +0x20
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details.(*Details).notifyObserver()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details/details.go:101 +0x2cc
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details.(*Details).RegisterObserver()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/upgrade/details/details.go:79 +0x150
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator.(*Coordinator).Upgrade()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator/coordinator.go:457 +0x640
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers.(*Upgrade).Handle.func1()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go:78 +0x178
[2023-10-13T13:08:03.453Z] 
[2023-10-13T13:08:03.453Z] Goroutine 115 (running) created at:
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers.TestUpgradeHandler()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers/handler_action_upgrade_test.go:75 +0x204
[2023-10-13T13:08:03.453Z]   testing.tRunner()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/.gvm/versions/go1.20.9.linux.arm64/src/testing/testing.go:1576 +0x180
[2023-10-13T13:08:03.453Z]   testing.(*T).Run.func1()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/.gvm/versions/go1.20.9.linux.arm64/src/testing/testing.go:1629 +0x40
[2023-10-13T13:08:03.453Z] 
[2023-10-13T13:08:03.453Z] Goroutine 117 (running) created at:
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers.(*Upgrade).Handle()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go:76 +0x760
[2023-10-13T13:08:03.453Z]   github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers.TestUpgradeHandler()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/src/github.com/elastic/elastic-agent/internal/pkg/agent/application/actions/handlers/handler_action_upgrade_test.go:80 +0x39c
[2023-10-13T13:08:03.453Z]   testing.tRunner()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/.gvm/versions/go1.20.9.linux.arm64/src/testing/testing.go:1576 +0x180
[2023-10-13T13:08:03.453Z]   testing.(*T).Run.func1()
[2023-10-13T13:08:03.453Z]       /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3527/.gvm/versions/go1.20.9.linux.arm64/src/testing/testing.go:1629 +0x40
[2023-10-13T13:08:03.453Z] ==================
```

Reply:
This flag shouldn't be an atomic, since it is exclusively used by the main Coordinator run loop. However, to clarify this kind of issue in the future, maybe add a comment to the effect of "If you get a race detector error while accessing this field, it probably means you're calling internal Coordinator methods from outside the Coordinator goroutine" :-)

Reply:
Agreed, when I saw this became atomic I started looking at the usages of it to make sure we were using it correctly (stores that should be compare and swap, etc). 

Then I realized the race was coming from the tests, I don't think it's worth the mental overhead of an atomic just for that.

Reply:
I wasn't smart enough to find the real source of the problem though :)